File: vibe-base/cpp/vibe-base.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <sstream>

struct State {
    std::size_t sales_table_index = 0;
    long sales_table_row = 0;
    long sales = 0;

    std::size_t expenses_table_index = 0;
    long expenses_table_row = 0;
    long expenses = 0;

    std::size_t number_of_tables = 0;

};

State state;

std::vector<std::string> split(const std::string& line, const char separator) {
    std::vector<std::string> parts;
    std::stringstream ss(line);
    std::string item;

    while (std::getline(ss, item, separator)) {
        parts.push_back(item);
    }
    return parts;
}

void process_header(const std::vector<std::string> fields, const std::size_t lineNumber) {
    state.number_of_tables++;

    if (fields[0] == "sales") {
        state.sales_table_index = std::stol(fields[fields.size() - 1]);
        auto tables = split(fields[1], ',');
        auto it = std::find(tables.begin(), tables.end(), "sale_value");
        if (it != tables.end()) {
            state.sales_table_row = std::distance(tables.begin(), it);
        }
    }

    if (fields[0] == "expenses") {
        state.expenses_table_index = std::stol(fields[fields.size() - 1]);
        auto tables = split(fields[1], ',');
        auto it = std::find(tables.begin(), tables.end(), "expense_value");
        if (it != tables.end()) {
            state.expenses_table_row = std::distance(tables.begin(), it);
        }
    }
}

void process_table(const std::string& line, const std::size_t lineNumber) {
    if ((lineNumber - state.number_of_tables) % state.number_of_tables == state.sales_table_index) {
        auto fields = split(line, ';');
        state.sales += std::stol(fields[state.sales_table_row]);;
    }

    if ((lineNumber - state.number_of_tables) % state.number_of_tables == state.expenses_table_index) {
        auto fields = split(line, ';');
        state.expenses += std::stol(fields[state.expenses_table_row]);
    }

}

void process(const std::string& line, const std::size_t lineNumber) {
    auto fields = split(line, '|');

    if (fields.size() == 3) {
        process_header(fields, lineNumber);
    } else {
        process_table(line, lineNumber);
    }

}

int main(int argc, const char * argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: vibe-base <filename>\n";
        return 1;
    }

    const char* filename = argv[1];
    std::ifstream in(filename);

    if (!in) {
        std::cerr << "Failed to open: " << filename << '\n';
        return 1;
    }

    std::string line;
    std::size_t lineNumber = 0;
    while (std::getline(in, line)) {
        if (line.empty()) {
            ++lineNumber;
            continue;
        }
        process(line, lineNumber);
        ++lineNumber;
    }

    std::cout << state.sales - state.expenses << '\n';
    return 0;
}

File: vibe-base/elixir/vibe-base.exs
defmodule VibeBase do
  def run([filename]) do
    File.stream!(filename)
    |> Stream.map(&String.trim/1)
    |> Enum.with_index()
    |> Enum.reduce(%{}, &process_line/2)
    |> subtract()
    |> IO.puts()
  end

  def run(_) do
    IO.puts("Usage: elixir vibe-base.exs <filename>")
  end

  defp process_line({line, _}, acc) when line == "" do
    acc
  end

  defp process_line({line, line_num}, acc) do
    parts =
      line
      |> String.split("|")
      |> Enum.map(&String.trim/1)

    case length(parts) do
      3 ->
        [table_name, column_list, start_row] = parts

        columns =
          column_list
          |> String.split(",")
          |> Enum.map(&String.trim/1)

        new_acc =
          acc
          |> Map.update(:number_of_tables, 1, &(&1 + 1))

        case table_name do
          "sales" ->
            new_acc
            |> Map.put(:sales, %{
              column_index: Enum.find_index(columns, &(&1 == "sale_value")),
              row_index: String.to_integer(start_row)
            })

          "expenses" ->
            new_acc
            |> Map.put(:expenses, %{
              column_index: Enum.find_index(columns, &(&1 == "expense_value")),
              row_index: String.to_integer(start_row)
            })

          _ ->
            new_acc
        end

      _ ->
        expenses = acc[:expenses]
        sales = acc[:sales]
        number_of_tables = acc[:number_of_tables]

        cond do
          sales && number_of_tables &&
              rem(line_num - number_of_tables, number_of_tables) == sales.row_index ->
            case Enum.at(String.split(line, ";"), sales.column_index) do
              value_str ->
                value = String.to_integer(value_str)
                Map.update(acc, "sale_value", value, &(&1 + value))
            end

          expenses && number_of_tables &&
              rem(line_num - number_of_tables, number_of_tables) == expenses.row_index ->
            case Enum.at(String.split(line, ";"), expenses.column_index) do
              value_str ->
                value = String.to_integer(value_str)
                Map.update(acc, "expense_value", value, &(&1 + value))
            end

          true ->
            acc
        end
    end
  end

  defp subtract(acc) do
    sales = Map.get(acc, "sale_value", 0)
    expenses = Map.get(acc, "expense_value", 0)
    sales - expenses
  end
end

VibeBase.run(System.argv())
